src/gui/properties/piecesbar.cpp

```diff
@@ -0,0 +1,305 @@
+    {
+    public:
+        PieceIndexToImagePos(const BitTorrent::TorrentHandle * const torrent, const QImage& image)
+            : m_piecesCount(torrent->info().piecesCount())
+            , m_imageWidth(image.width())
+        {
+        }
+
+        int imagePos(int pieceIndex) const
+        {
+            return static_cast<float>(pieceIndex) / (m_piecesCount - 1) * (m_imageWidth - 1);
+        }
+
+        int pieceIndex(int imagePos) const
+        {
+            return static_cast<float>(imagePos) / (m_imageWidth - 1) * (m_piecesCount - 1);
+        }
+
+    private:
+        const int m_piecesCount;
+        const int m_imageWidth;
+    };
+}
+
+PiecesBar::PiecesBar(bool local, QWidget *parent)
+    : QWidget(parent)
+    , m_local{local}
+    , m_torrent{nullptr}
+    , m_image{}
+    , m_borderColor{palette().color(QPalette::Dark)}
+    , m_bgColor{Qt::white}
+    , m_pieceColor{Qt::blue}
+    , m_pieceColors{}
+    , m_hovered{false}
+    , m_highlitedRegion{}
+{
+    updatePieceColors();
+    setMouseTracking(true);
+}
+
+void PiecesBar::setTorrent(BitTorrent::TorrentHandle *torrent)
+{
+    m_torrent = torrent;
+    if (!m_torrent)
+        clear();
+}
+
+void PiecesBar::clear()
+{
+    m_image = QImage();
+    update();
+}
+
+void PiecesBar::setColors(const QColor &background, const QColor &border, const QColor &complete)
+{
+    m_bgColor = background;
+    m_borderColor = border;
+    m_pieceColor = complete;
+
+    updatePieceColors();
+    requestImageUpdate();
+}
+
+bool PiecesBar::event(QEvent *e)
+{
+    if (e->type() == QEvent::ToolTip) {
+        showToolTip(static_cast<QHelpEvent*>(e));
+        return true;
+    }
+    else {
+        return QWidget::event(e);
+    }
+}
+
+void PiecesBar::enterEvent(QEvent *e)
+{
+    m_hovered = true;
+    base::enterEvent(e);
+}
+
+void PiecesBar::leaveEvent(QEvent *e)
+{
+    m_hovered = false;
+    m_highlitedRegion = QRect();
+    requestImageUpdate();
+    base::leaveEvent(e);
+}
+
+void PiecesBar::mouseMoveEvent(QMouseEvent *e)
+{
+    // if user pointed to a piece which is a part of a single large file,
+    // we highlight the space, occupied by this file
+    highlightFile(e->pos().x() - borderWidth);
+    base::mouseMoveEvent(e);
+}
+
+void PiecesBar::paintEvent(QPaintEvent *)
+{
+    QPainter painter(this);
+    QRect imageRect(borderWidth, borderWidth, width() - 2 * borderWidth, height() - 2 * borderWidth);
+    if (m_image.isNull()) {
+        painter.setBrush(Qt::white);
+        painter.drawRect(imageRect);
+    }
+    else {
+        if (m_image.width() != imageRect.width())
+            updateImage(m_image);
+        painter.drawImage(imageRect, m_image);
+    }
+
+    if (!m_highlitedRegion.isNull()) {
+        QColor highlightColor {this->palette().color(QPalette::Active, QPalette::Highlight)};
+        highlightColor.setAlphaF(0.35);
+        QRect targetHighlightRect {m_highlitedRegion.adjusted(borderWidth, borderWidth, borderWidth, height() - 2 * borderWidth)};
```

Comment:
Fix coding style!
Brace enclosed initializers are allowed in list or aggregate contexts only. Not for copy constructors.